lib/drivers/glucorx/glucoRxDriver.js

```diff
@@ -0,0 +1,632 @@
+    bytes[3] = dateTime.hours & 0x1F;
+
+    const result = await this.commandResponse(COMMAND.WRITE_TIME, bytes);
+    const newDateTime = GlucoRx.parseDateTime(result);
+    newDateTime.year -= 2000;
+
+    if (!_.isEqual(dateTime, newDateTime)) {
+      debug('Set date/time:', dateTime);
+      debug('Received date/time:', newDateTime);
+      throw new Error('Error setting date/time.');
+    }
+  }
+
+  static probe(cb) {
+    debug('not probing GlucoRx');
+    cb();
+  }
+
+  async ping() {
+    const bytesWritten = await this.hidDevice.sendPromisified(this.buildPacket(COMMAND.READ_MODEL));
+    debug('Sent', bytesWritten, 'bytes.');
+    const result = await this.hidDevice.receiveTimeout(READ_TIMEOUT);
+    debug('Received:', common.bytes2hex(result));
+
+    if (result.length === 0) {
+      if (this.retries <= 3) {
+        debug('Retrying..');
+        this.retries += 1;
+        await this.ping();
+      } else {
+        throw new Error('Device not responding.');
+      }
+    } else {
+      this.retries = 0;
+    }
+  }
+}
+
+module.exports = (config) => {
+  const cfg = _.clone(config);
+  _.assign(cfg.deviceInfo, {
+    tags: ['bgm'],
+    manufacturers: ['GlucoRx'],
+  });
+
+  let hidDevice = config.deviceComms;
+
+  const driver = new GlucoRx(cfg);
+
+  const hasFlag = (flag, v) => {
+    // eslint-disable-next-line no-bitwise
+    if ((flag.value & v) === flag.value) {
+      return true;
+    }
+    return false;
+  };
+
+  const buildBGRecords = (data) => {
+    _.forEach(data.records, (record, index) => {
+      let { value } = record;
+
+      // According to spec, HI > 600 and LO < 20
+      let annotation = null;
+      if (value > 600) {
+        value = 601;
+        annotation = {
+          code: 'bg/out-of-range',
+          value: 'high',
+          threshold: 600,
+        };
+      } else if (value < 20) {
+        value = 19;
+        annotation = {
+          code: 'bg/out-of-range',
+          value: 'low',
+          threshold: 20,
+        };
+      } else {
+        value = _.toInteger(value);
+      }
+
+      if (!hasFlag(FLAGS.CONTROL_SOLUTION, record.flags)
+          && !hasFlag(FLAGS.KETONE, record.flags)
+          && !hasFlag(FLAGS.HCT, record.flags)) {
+        const recordBuilder = cfg.builder.makeSMBG()
+          .with_value(value)
+          .with_units('mg/dL') // values are always in 'mg/dL'
+          .with_deviceTime(sundial.formatDeviceTime(record.jsDate))
+          .set('index', index);
+
+        cfg.tzoUtil.fillInUTCInfo(recordBuilder, record.jsDate);
+
+        if (annotation) {
+          annotate.annotateEvent(recordBuilder, annotation);
+        }
+
+        const postRecord = recordBuilder.done();
+        delete postRecord.index;
+        data.post_records.push(postRecord);
+      }
+    });
+  };
+
+  const buildKetoneRecords = (data) => {
+    _.forEach(data.records, (record, index) => {
+      if (hasFlag(FLAGS.KETONE, record.flags)) {
+        let value = record.value / 30.0;
```

Comment:
Is this magic `30.0` constant just a part of the spec?

Reply:
Yep, for some reason they decided to be able to store mmol/L values as integers they just multiply the value by 30 :shrug: